packages/contracts-periphery/contracts/universal/op-nft/OptimistInviter.sol

```diff
@@ -0,0 +1,234 @@
+        0x6529fd129351e725d7bcbc468b0b0b4675477e56b58514e69ab7e66ddfd20fce;
+
+    /**
+     * @notice Attestation key for granting invites.
+     *         bytes32("optimist.can-invite")
+     */
+    bytes32 public constant CAN_INVITE_ATTESTATION_KEY =
+        0x6f7074696d6973742e63616e2d696e7669746500000000000000000000000000;
+
+    /**
+     * @notice Attestation key allowing the attested account to mint.
+     *         bytes32("optimist.can-mint-from-invite")
+     */
+    bytes32 public constant CAN_MINT_FROM_INVITE_ATTESTATION_KEY =
+        0x6f7074696d6973742e63616e2d6d696e742d66726f6d2d696e76697465000000;
+
+    /**
+     * @notice Granter who can set accounts' invite counts.
+     */
+    address public immutable INVITE_GRANTER;
+
+    /**
+     * @notice Address of the AttestationStation contract.
+     */
+    AttestationStation public immutable ATTESTATION_STATION;
+
+    /**
+     * @notice Struct that represents a claimable invite that will be signed by the issuer.
+     *
+     * @custom:field issuer   Address that issued the signature. Reason this is explicitly included,
+     *                        and not implicitly assumed to be the recovered address from the
+     *                        signature is that the issuer may be using a ERC-1271 compatible
+     *                        contract wallet, where the recovered address is not the same as the
+     *                        issuer, or the signature is not an ECDSA signature at all.
+     * @custom:field nonce    Pseudorandom nonce to prevent replay attacks.
+     */
+    struct ClaimableInvite {
+        address issuer;
+        bytes32 nonce;
+    }
+
+    /**
+     * @notice Maps from hashes to whether or not they have been committed.
+     */
+    mapping(bytes32 => bool) public commitments;
+
+    /**
+     * @notice Maps from addresses to nonces to whether or not they have been used.
+     */
+    mapping(address => mapping(bytes32 => bool)) public usedNonces;
+
+    /**
+     * @custom:semver 1.0.0
+     *
+     * @param _inviteGranter      Address of the invite granter.
+     * @param _attestationStation Address of the AttestationStation contract.
+     */
+    constructor(address _inviteGranter, AttestationStation _attestationStation) Semver(1, 0, 0) {
+        INVITE_GRANTER = _inviteGranter;
+        ATTESTATION_STATION = _attestationStation;
+    }
+
+    /**
+     * @notice Initializes the OptimistInviter contract, setting the EIP712 context.
+     *
+     * @param _name Contract name
+     */
+    function initialize(string memory _name) public initializer {
+        __EIP712_init(_name, version());
+    }
+
+    /**
+     * @notice Allows invite granter to set the number of invites an address has.
+     *
+     * @param _accounts    An array of accounts to update the invite counts of.
+     * @param _inviteCount Number of invites to set to.
+     */
+    function setInviteCounts(address[] calldata _accounts, uint256 _inviteCount) public {
```

Comment:
I would just name this attest and same api as attestation station

Reply:
Steel man for special implementation:
- no need to write js code turning invites to attestations
- No worry about a junky Attestation causing issues later with say an indexer
- Better etherscan ui
- Self documenting easier to understand how to use contract

Steel man for polymorphism
- all tooling like atst sdk just plugs right in
- Less lines of solidity code
- Slightly less gas
- Simpler to understand just following a simple proxy design pattern
- maximally flexible.   You can issue different invite counts per account for example

I’m partial for the second because in general everything is much easier the further up the stack you go if things have polymorphism with their interfaces and it is more flexible

Reply:
While I’m aligned for the need for a more general AttestationStation proxy contract, I’m leaning towards keeping the current implementation, and keep the scope of this contract as minimal as possible.
My main point of concern is just that I don’t think we should make this contract a full proxy for the multi-sig, that allows the owner to make any arbitrary attestations through this contract.
This is specifically a contract that is meant to handle issuing invites, and the goal wasn’t to make a general purpose ATST proxy our multi-sig. I feel that having an attest function that allows any type of attestation to be made will hurt readability of the contract.
Having this contract only be concerned with issuing a very limited set of attestations, “optimist.can-invite” and “optimist.can-mint-from-invite” makes it clear how this contract is meant to be used, and what this contract should or shouldn’t be able to do.